store/localstore/local_region.go

```diff
@@ -0,0 +1,318 @@
+package localstore
+
+import (
+	"bytes"
+
+	"encoding/binary"
+	"github.com/golang/protobuf/proto"
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/terror"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+	"github.com/pingcap/tidb/xapi/tablecodec"
+	"github.com/pingcap/tidb/xapi/tipb"
+)
+
+// local region server.
+type localRegion struct {
+	id       int
+	store    *dbStore
+	startKey []byte
+	endKey   []byte
+}
+
+type regionRequest struct {
+	Tp       int64
+	data     []byte
+	startKey []byte
+	endKey   []byte
+}
+
+type regionResponse struct {
+	req  *regionRequest
+	err  error
+	data []byte
+	// If region missed some request key range, newStartKey and newEndKey is returned.
+	newStartKey []byte
+	newEndKey   []byte
+}
+
+func (rs *localRegion) Handle(req *regionRequest) (*regionResponse, error) {
+	resp := &regionResponse{
+		req: req,
+	}
+	if req.Tp == kv.ReqTypeSelect || req.Tp == kv.ReqTypeIndex {
+		sel := new(tipb.SelectRequest)
+		err := proto.Unmarshal(req.data, sel)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		txn := newTxn(rs.store, kv.Version{Ver: uint64(*sel.StartTs)})
+		var rows []*tipb.Row
+		if req.Tp == kv.ReqTypeSelect {
+			rows, err = rs.getRowsFromSelectReq(txn, sel)
+		} else {
+			rows, err = rs.getRowsFromIndexReq(txn, sel)
+		}
+		selResp := new(tipb.SelectResponse)
+		selResp.Error = toPBError(err)
+		selResp.Rows = rows
+		resp.err = err
+		data, err := proto.Marshal(selResp)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		resp.data = data
+	}
+	if bytes.Compare(rs.startKey, req.startKey) < 0 || bytes.Compare(rs.endKey, req.endKey) > 0 {
+		resp.newStartKey = rs.startKey
+		resp.newEndKey = rs.endKey
+	}
+	return resp, nil
+}
+
+func (rs *localRegion) getRowsFromSelectReq(txn kv.Transaction, sel *tipb.SelectRequest) ([]*tipb.Row, error) {
+	tid := sel.TableInfo.GetTableId()
+	kvRanges := rs.extractRecordKVRanges(tid, sel.Ranges)
+	var handles []int64
+	for _, ran := range kvRanges {
+		ranHandles, err := seekRangeHandles(tid, txn, ran)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		handles = append(handles, ranHandles...)
+	}
+	var rows []*tipb.Row
+	for _, handle := range handles {
+		row, err := rs.getRowByHandle(txn, tid, handle, sel.TableInfo.Columns)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		rows = append(rows, row)
+	}
+	return rows, nil
+}
+
+func (rs *localRegion) extractRecordKVRanges(tid int64, krans []*tipb.KeyRange) []kv.KeyRange {
+	var kvRanges []kv.KeyRange
+	for _, kran := range krans {
+		upperKey := tablecodec.EncodeRowKey(tid, kran.GetHigh())
+		if bytes.Compare(upperKey, rs.startKey) <= 0 {
```

Comment:
`extractRecordKVRanges` and `extractIndexKVRanges` are just different encoding, could be unified?